### PR TITLE
Rails 5.2 upgrade

### LINKED ISSUE #7093: Upgrade to Rails 5.2

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #7093

#### What should we test?
Everything :-)

#### Release notes
Changelog Category: Technical changes
Rails upgrade to v5.2

### CODE REVIEW

**Comment by mkllnk** [comment]
This looks like work in progress, moving back to dev.

**Comment by Matt-Yorkley** [comment]
Next steps:

- pull as many of the changes as we can into PRs to master
- tidy up the branch and get ti ready for review
- possibly add some additional unit tests around those places where we updated the guard clauses in after_save callbacks?

File: app/controllers/spree/orders_controller.rb [hunk 1]
```diff
@@ -104,7 +104,7 @@ def update
       else
         # Show order with original values, not newly entered ones
         @insufficient_stock_lines = @order.insufficient_stock_lines
-        @order.line_items(true)
+        @order.line_items.reload
```
Comment (luisramos0) [inline app/controllers/spree/orders_controller.rb:107]: can go to master.

File: app/helpers/spree/admin/orders_helper.rb [hunk 2]
```diff
@@ -50,13 +50,13 @@ def ticket_links
 
       def edit_order_link
         { name: t(:edit_order),
-          url: edit_admin_order_path(@order),
+          url: spree.edit_admin_order_path(@order),
```
Comment (luisramos0) [inline app/helpers/spree/admin/orders_helper.rb:53]: can go to master.

File: app/helpers/spree/payment_methods_helper.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module Spree
+  module PaymentMethodsHelper
```
Comment (luisramos0) [inline app/helpers/spree/payment_methods_helper.rb:2]: can go to master.

File: app/mailers/spree/order_mailer.rb [hunk 4]
```diff
@@ -4,7 +4,7 @@ module Spree
   class OrderMailer < BaseMailer
     helper 'checkout'
     helper SpreeCurrencyHelper
-    helper Spree::Admin::PaymentsHelper
+    helper Spree::PaymentMethodsHelper
```
Comment (luisramos0) [inline app/mailers/spree/order_mailer.rb:7]: can go to master.

File: app/models/customer.rb [hunk 5]
```diff
@@ -46,6 +46,6 @@ def check_for_orders
     return true unless orders.any?
 
     errors[:base] << I18n.t('admin.customers.destroy.has_associated_orders')
-    false
+    throw :abort
```
Comment (luisramos0) [inline app/models/customer.rb:49]: can go to master.

File: app/models/spree/payment.rb [hunk 6]
```diff
@@ -201,7 +201,7 @@ def invalidate_old_payments
     end
 
     def update_order
-      order.reload.update!
+      order.update!
```
Comment (luisramos0) [inline app/models/spree/payment.rb:204]: can go to master.

File: app/services/cart_service.rb [hunk 7]
```diff
@@ -106,13 +106,15 @@ def read_variants(data)
   end
 
   def read_variants_hash(data)
-    (data[:variants] || []).map do |variant_id, quantity|
-      if quantity.is_a?(Hash)
-        { variant_id: variant_id, quantity: quantity[:quantity], max_quantity: quantity[:max_quantity] }
+    variants_array = []
+    (data[:variants] || []).each do |variant_id, quantity|
+      if quantity.is_a?(ActionController::Parameters)
```
Comment (luisramos0) [inline app/services/cart_service.rb:111]: can go to master.

File: app/services/permitted_attributes/enterprise.rb [hunk 8]
```diff
@@ -7,7 +7,7 @@ def initialize(params)
     end
 
     def call
-      return @params[:enterprise] if @params[:enterprise].blank?
+      return {} if @params[:enterprise].blank?
```
Comment (luisramos0) [inline app/services/permitted_attributes/enterprise.rb:10]: can go to master.

File: config/application.rb [hunk 9]
```diff
@@ -79,7 +79,8 @@ class Application < Rails::Application
     initializer 'ofn.spree_locale_settings', before: 'spree.promo.environment' do |app|
       Spree::Config['checkout_zone'] = ENV['CHECKOUT_ZONE']
       Spree::Config['currency'] = ENV['CURRENCY']
-      if Spree::Country.table_exists?
+
+      if ActiveRecord::Base.connected? && Spree::Country.table_exists?
```
Comment (luisramos0) [inline config/application.rb:83]: can go to master.

File: db/migrate/20181128054803_old_migrations_removed.rb [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
-class OldMigrationsRemoved < ActiveRecord::Migration
+class OldMigrationsRemoved < ActiveRecord::Migration[4.2]
```
Comment (luisramos0) [inline db/migrate/20181128054803_old_migrations_removed.rb:3]: can go to master.

File: lib/spree/core/controller_helpers/order.rb [hunk 11]
```diff
@@ -52,7 +52,7 @@ def spree_current_order(create_order_if_necessary = false)
 
           return unless @current_order
 
-          @current_order.last_ip_address = ip_address
+          @current_order.update_columns(last_ip_address: ip_address)
```
Comment (luisramos0) [inline lib/spree/core/controller_helpers/order.rb:55]: can go to master.

File: lib/spree/localized_number.rb [hunk 12]
```diff
@@ -66,5 +68,15 @@ def self.add_trailing_zeros(number)
       # If does not end in ,00 / .00 then add trailing 00 to turn it into cents
       number << "00" unless number =~ /^.*[\.,]\d{2}$/
     end
+
+    private
+
+    def non_activerecord_attribute?(attribute)
```
Comment (luisramos0) [inline lib/spree/localized_number.rb:74]: can go to master.

File: spec/controllers/api/v0/orders_controller_spec.rb [hunk 13]
```diff
@@ -274,7 +274,7 @@ module Api
 
       before do
         order.finalize!
-        create(:check_payment, order: order, amount: order.total)
+        order.payments << create(:check_payment, order: order, amount: order.total)
```
Comment (luisramos0) [inline spec/controllers/api/v0/orders_controller_spec.rb:277]: can go to master.

File: spec/controllers/checkout_controller_spec.rb [hunk 14]
```diff
@@ -46,7 +46,7 @@
     before do
       allow(controller).to receive(:current_order).and_return(order)
       allow(order).to receive(:distributor).and_return(distributor)
-      order.order_cycle = order_cycle
+      order.update(order_cycle: order_cycle)
```
Comment (luisramos0) [inline spec/controllers/checkout_controller_spec.rb:49]: can go to master.

File: spec/features/admin/orders_spec.rb [hunk 15]
```diff
@@ -65,7 +65,7 @@
     context "with a capturable order" do
       before do
         order.finalize! # ensure order has a payment to capture
-        create :check_payment, order: order, amount: order.total
+        order.payments << create(:check_payment, order: order, amount: order.total)
```
Comment (luisramos0) [inline spec/features/admin/orders_spec.rb:68]: can go to master.

File: spec/features/consumer/shopping/checkout_stripe_spec.rb [hunk 16]
```diff
@@ -219,7 +219,8 @@
           stub_add_metadata_request(payment_method: "pm_123", response: {})
           stub_payment_intents_post_request order: order
           stub_successful_capture_request order: order
-          stub_customers_post_request email: user.email
+          stub_customers_post_request email: "test@test.com" # First checkout with default details
```
Comment (luisramos0) [inline spec/features/consumer/shopping/checkout_stripe_spec.rb:222]: can go to master.

File: spec/models/spree/classification_spec.rb [hunk 17]
```diff
@@ -4,14 +4,16 @@
 
 module Spree
   describe Classification do
-    let(:product) { build_stubbed(:simple_product) }
-    let(:taxon) { build_stubbed(:taxon) }
+    let(:product) { create(:simple_product) }
```
Comment (luisramos0) [inline spec/models/spree/classification_spec.rb:7]: can go to master.

File: spec/views/spree/admin/orders/edit.html.haml_spec.rb [hunk 18]
```diff
@@ -4,6 +4,9 @@
 
 describe "spree/admin/orders/edit.html.haml" do
   helper Spree::BaseHelper # required to make pretty_time work
+  helper Spree::Admin::NavigationHelper
```
Comment (luisramos0) [inline spec/views/spree/admin/orders/edit.html.haml_spec.rb:7]: can go to master.

**Review by luisramos0 — COMMENTED** [review]
Yeah, more than half of this PR can go to master 👍 
i left review comments, maybe we can use these as a TODO list.

**Comment by andrewpbrett** [comment]
I started on separating things here: https://github.com/openfoodfoundation/openfoodnetwork/pull/7413

**Comment by luisramos0** [comment]
Great Andy, thanks!
This is good to be tested :+1:  
I cant approve because I created the PR but anyway :heavy_check_mark:

File: Gemfile.lock [hunk 19]
```diff
@@ -118,7 +122,7 @@ GEM
       railties (>= 3, < 7)
     angularjs-file-upload-rails (2.4.1)
     angularjs-rails (1.5.5)
-    arel (7.1.4)
+    arel (9.0.0)
```
Comment (sauloperez) [inline Gemfile.lock:125]: :heart_eyes: I think this bump comes with lots of goodies

**Review by sauloperez — APPROVED** [review]

**Review by Matt-Yorkley — APPROVED** [review]
:tada:

**Comment by filipefurtad0** [comment]
Hi! I just noticed the merge conflicts. Should we address them before testing?

**Comment by Matt-Yorkley** [comment]
Rebased :+1:

**Comment by Matt-Yorkley** [comment]
FYI; I just rebased on master and had to re-apply a commit related to https://github.com/openfoodfoundation/openfoodnetwork/issues/7494 :+1:

**Comment by filipefurtad0** [comment]
This is ready to go!
Only issue found was #7942 and it was addressed :+1: 

Failing spec should be flaky as discussed with @Matt-Yorkley (re-build pending).

Testing notes here:
https://docs.google.com/document/d/1vJp95Rh2pd7vYIvW0F8-63oAbTbgbCxLz14NnlLFtRc/edit#heading=h.4ctypvfojpgm

Moving to Ready to go :tada:

**Pull request merged**